src/ast/analysis/Type.cpp

```diff
@@ -823,23 +840,282 @@ void TypeAnalysis::print(std::ostream& os) const {
     }
 }
 
+TypeAttribute TypeAnalysis::getFunctorReturnType(const Functor* functor) const {
+    assert(hasProcessedFunctor(functor) && "functor not yet processed");
+    if (auto* intrinsic = as<IntrinsicFunctor>(functor)) {
+        return functorInfo.at(intrinsic)->result;
+    } else if (const auto* udf = as<UserDefinedFunctor>(functor)) {
+        return udfDeclaration.at(udf->getName())->getReturnType();
+    }
+    fatal("Missing functor type.");
+}
+
+TypeAttribute TypeAnalysis::getFunctorArgType(const Functor* functor, const size_t idx) const {
+    assert(hasProcessedFunctor(functor) && "functor not yet processed");
+    if (auto* intrinsic = as<IntrinsicFunctor>(functor)) {
+        auto* info = functorInfo.at(intrinsic);
+        return info->params.at(info->variadic ? 0 : idx);
+    } else if (auto* udf = as<UserDefinedFunctor>(functor)) {
+        return udfDeclaration.at(udf->getName())->getArgsTypes().at(idx);
+    }
+    fatal("Missing functor type.");
+}
+
+const std::vector<TypeAttribute>& TypeAnalysis::getFunctorArgTypes(const UserDefinedFunctor& udf) const {
+    return udfDeclaration.at(udf.getName())->getArgsTypes();
+}
+
+bool TypeAnalysis::isStatefulFunctor(const UserDefinedFunctor* udf) const {
+    return udfDeclaration.at(udf->getName())->isStateful();
+}
+
+const std::map<const NumericConstant*, NumericConstant::Type>& TypeAnalysis::getNumericConstantTypes() const {
+    return numericConstantType;
+}
+
+bool TypeAnalysis::isMultiResultFunctor(const Functor& functor) {
+    if (isA<UserDefinedFunctor>(functor)) {
+        return false;
+    } else if (auto* intrinsic = as<IntrinsicFunctor>(functor)) {
+        auto candidates = functorBuiltIn(intrinsic->getBaseFunctionOp());
+        assert(!candidates.empty() && "at least one op should match");
+        return candidates[0].get().multipleResults;
+    }
+    fatal("Missing functor type.");
+}
+
+IntrinsicFunctors TypeAnalysis::validOverloads(const IntrinsicFunctor& func) const {
+    auto typeAttrs = [&](const Argument* arg) -> std::set<TypeAttribute> {
+        std::set<TypeAttribute> tyAttrs;
+        if (const auto* inf = dynamic_cast<const IntrinsicFunctor*>(arg)) {
+            if (hasProcessedFunctor(inf)) {
+                tyAttrs.insert(getFunctorReturnType(inf));
+                return tyAttrs;
+            }
+        }
+        auto&& types = getTypes(arg);
+        if (types.isAll())
+            return {TypeAttribute::Signed, TypeAttribute::Unsigned, TypeAttribute::Float,
+                    TypeAttribute::Symbol, TypeAttribute::Record};
+
+        for (auto&& ty : types)
```

Comment:
For two characters more you could have a meaningful name that matches the container.